Step execution for Kore - DO NOT REVIEW YET

virgil-serbanuta opened the pull request.

bmmoore commented:
Topological sorting is also reinventing the wheel, `Data.Graph.stronglyConnComp` will return a topological sort wrapped in `AcyclicSCC`  if the graph is acyclic (a node with an edge to itself will end up in single-element `CyclicSCC` if it isn't in a larger component).

virgil-serbanuta commented:
@u-- @bmmoore Thank you both, I didn't know about these.

Data.Graph is actually kind of weird, so I'm not sure I'd trust whoever built it to do something reasonable. I mean, topSort, which is documented as sorting the graph topologically,  does not do a topological sort (since it does not fail, or at least notify the caller, when it's impossible to sort topologically), but it seems that it's easy to do a proper topological sorting based on the output of stronglyConnComp. Really?

Anyway, since @bmmoore reccomends it, I'll switch to stronglyConnComp and hope that their implementation does not match their interface design.

virgil-serbanuta commented:
Actually, I just realized that stronglyConnComp may not be what we need. I keep hearing that, in some vague future, we'll have to do do proofs in which we have to explain why we were unable to apply an axiom (or something like that). @bmmoore Do you know if we'll need an actual cycle for that? If yes, then stronglyConnComp may not be good enough because CyclicSCC does not provide a cycle example (although all of its nodes are part of cycles).

virgil-serbanuta commented:
Since @traiansf wanted some progress, I stopped waiting for an answer from @bmmoore and switched to the Data.Graph functions. I also removed SetWithInsertionOrder.

traiansf commented on src/main/haskell/language-kore/test/parser/Data/Kore/Step/BaseStepTest.hs:
```diff
@@ -0,0 +1,777 @@
+            )
+        -- sigma(sigma(x, x), sigma(y, y)) -> sigma(x, y)
+        -- vs
+        -- sigma(sigma(a, f(b)), sigma(b, a))
```
maybe you should consider adding the expected result of the one step, too.  I know it should be obvious, but still...

virgil-serbanuta replied:
Done.

bmmoore commented:
I'm not sure what all it will be used for, but if topological sorting is for a well-formedness check like looking for cyclic imports then we don't need to exhibit a path. If we do, `fgl` is an established graph library with a basic interface similar to `Data.Graph` but much more available functionality (the advantage of `Data.Graph` is that it's in the `containers` package). If we actually need one of the fancy data structure for incrementally extending a graph while checking whether it stays acyclic then we would need to implement it ourself.

virgil-serbanuta commented on src/main/haskell/language-kore/test/parser/Data/Kore/Comparators.hs:
```diff
@@ -0,0 +1,672 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```
Better now?

virgil-serbanuta commented on src/main/haskell/language-kore/test/parser/Data/Kore/Algorithm/TopologicalSortTest.hs:
```diff
@@ -4,10 +4,11 @@ module Data.Kore.Algorithm.TopologicalSortTest
 import           Test.Tasty                          (TestTree, testGroup)
 import           Test.Tasty.HUnit                    (assertEqual, testCase)
 
+import           Control.Applicative                 ((<|>))
```
Removed.
